packages/Discretization/src/DTK_Interpolation_def.hpp

```diff
@@ -0,0 +1,669 @@
+                }
+            } );
+        Kokkos::fence();
+    }
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::performDistributedSearch(
+    Kokkos::View<double **, DeviceType> points_coord,
+    Kokkos::View<DataTransferKit::Point *, DeviceType> &imported_points,
+    Kokkos::View<int *, DeviceType> &imported_query_ids,
+    Kokkos::View<int *, DeviceType> &imported_cell_indices )
+{
+    DataTransferKit::DistributedSearchTree<DeviceType> distributed_tree(
+        _comm, _bounding_boxes );
+
+    unsigned int const n_points = points_coord.extent( 0 );
+
+    // Build the queries
+    // FIXME do not use Within predicate because it requires a radius, i.e., it
+    // is mesh dependent
+    using ExecutionSpace = typename DeviceType::execution_space;
+    Kokkos::View<DataTransferKit::Details::Within *, DeviceType> queries(
+        "queries", n_points );
+    Kokkos::parallel_for( "register_queries",
+                          Kokkos::RangePolicy<ExecutionSpace>( 0, n_points ),
+                          KOKKOS_LAMBDA( int i ) {
+                              queries( i ) = DataTransferKit::Details::within(
+                                  {{points_coord( i, 0 ), points_coord( i, 1 ),
+                                    points_coord( i, 2 )}},
+                                  1e-14 );
+                          } );
+    Kokkos::fence();
+
+    // Perform the distributed search
+    Kokkos::View<int *, DeviceType> indices( "indices" );
+    Kokkos::View<int *, DeviceType> offset( "offset" );
+    Kokkos::View<int *, DeviceType> ranks( "ranks" );
+    distributed_tree.query( queries, indices, offset, ranks );
+
+    // Create the distributor and its reverse that will be used to communicate
+    // the results back to the calling processors.
+    auto ranks_host = Kokkos::create_mirror_view( ranks );
+    Kokkos::deep_copy( ranks_host, ranks );
+    Tpetra::Distributor distributed_search_distributor( _comm );
+    unsigned int const n_imports =
+        distributed_search_distributor.createFromSends(
+            Teuchos::ArrayView<int const>( ranks_host.data(),
+                                           ranks_host.extent( 0 ) ) );
+
+    // Communicate cell indices
+    Kokkos::realloc( imported_cell_indices, n_imports );
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        distributed_search_distributor, indices, imported_cell_indices );
+    // Duplicate the points_coord for the communication. Duplicating the points
+    // allows us to use the same distributor.
+    unsigned int const indices_size = indices.extent( 0 );
+    Kokkos::View<DataTransferKit::Point *, DeviceType> exported_points(
+        "exported_points", indices_size );
+    Kokkos::View<int *, DeviceType> exported_query_ids( "exported_query_ids",
+                                                        indices_size );
+    if ( offset.extent( 0 ) > 0 )
```

Comment:
Can it actually be zero?  I would expect it is going to be of size n_queries + 1.  I remember testing (on node) for search on empty trees but not searching for nothing.  There is definitely no test in the distributed case.

Reply:
It can be zero. In one of the test, there is a processor that does not own any points and does not do any query.

Reply:
I need to look more into this.

Reply:
Per #305 you can get rid of it.